static-check/src/types/type_.rs

```diff
@@ -0,0 +1,330 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the Leo library.
+
+// The Leo library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The Leo library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the Leo library. If not, see <https://www.gnu.org/licenses/>.
+use crate::{SymbolTable, TypeError, TypeVariable};
+use leo_typed::{Identifier, IntegerType, Span, Type as UnresolvedType};
+
+use serde::{Deserialize, Serialize};
+use std::{
+    cmp::{Eq, PartialEq},
+    fmt,
+};
+
+/// A type in a Leo program.
+#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
```

Comment:
the hand-written `PartialEq` impl (which is a good thing) for this type renders the `Hash` derivation invalid: https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq; `Hash` needs to be implemented by hand too